pstoreds/addr_book.go

```diff
@@ -2,398 +2,338 @@ package pstoreds
+func (ab *dsAddrBook) Close() {
+	ab.cancelFn()
+	ab.done.Wait()
+}
 
-	if err != nil {
-		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
-		return err
+func (ab *dsAddrBook) asyncFlush(pr *addrsRecord) {
+	select {
+	case ab.flushJobCh <- pr:
+	default:
+		log.Warningf("flush queue is full; could not flush record for peer %s", pr.Id.ID)
 	}
+}
 
-	// Update was successful, so broadcast event only for new addresses.
-	for i, _ := range keys {
-		if !existed[i] {
-			mgr.subsManager.BroadcastAddr(p, addrs[i])
+// loadRecord is a read-through fetch. It fetches a record from cache, falling back to the
+// datastore upon a miss, and returning a newly initialized record if the peer doesn't exist.
+//
+// loadRecord calls clean() on existing recordsrecord before returning it. If the record changes
+// as a result and the update argument is true, an async flush is queued.
+//
+// If the cache argument is true, the record is inserted in the cache when loaded from the datastore.
+func (ab *dsAddrBook) loadRecord(id peer.ID, cache bool, update bool) (pr *addrsRecord, err error) {
+	if e, ok := ab.cache.Get(id); ok {
+		pr = e.(*addrsRecord)
+		if pr.clean() && update {
+			ab.asyncFlush(pr)
 		}
+		return pr, nil
 	}
-	return nil
-}
 
-// dbInsert performs a transactional insert of the provided keys and values.
-func (mgr *dsAddrBook) dbInsert(keys []ds.Key, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) ([]bool, error) {
-	var (
-		err     error
-		existed = make([]bool, len(keys))
-		exp     = time.Now().Add(ttl)
-	)
+	pr = &addrsRecord{AddrBookRecord: &pb.AddrBookRecord{}}
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(id)))
+	data, err := ab.ds.Get(key)
 
-	txn, err := mgr.ds.NewTransaction(false)
-	if err != nil {
+	switch err {
+	case ds.ErrNotFound:
+		pr.Id = &pb.ProtoPeerID{ID: id}
+	case nil:
+		if err = pr.Unmarshal(data); err != nil {
+			return nil, err
+		}
+		if pr.clean() && update {
+			ab.asyncFlush(pr)
+		}
+	default:
 		return nil, err
 	}
-	defer txn.Discard()
 
-	ttltxn := txn.(ds.TTLDatastore)
-	for i, key := range keys {
-		// Check if the key existed previously.
-		if existed[i], err = ttltxn.Has(key); err != nil {
-			log.Errorf("transaction failed and aborted while checking key existence: %s, cause: %v", key.String(), err)
-			return nil, err
-		}
+	if cache {
+		ab.cache.Add(id, pr)
+	}
+	return pr, nil
+}
 
-		// The key embeds a hash of the value, so if it existed, we can safely skip the insert and
-		// just update the TTL.
-		if existed[i] {
-			switch mode {
-			case ttlOverride:
-				err = ttltxn.SetTTL(key, ttl)
-			case ttlExtend:
-				var curr time.Time
-				if curr, err = ttltxn.GetExpiration(key); err == nil && exp.After(curr) {
-					err = ttltxn.SetTTL(key, ttl)
+// flusher is a goroutine that takes care of persisting asynchronous flushes to the datastore.
+func (ab *dsAddrBook) flusher() {
+	ab.done.Add(1)
```

Comment:
WaitGroup should be increased before goroutine creation. This is to prevent subtle races that will cause panics.

Example of such race, in this case, would be:
```go
addrb, _ := NewAddrBook(...)
addrb.Close()
```

As goroutine scheduling is not defined there is a chance that this `ab.done.Add(1)` statement is executed at the same time `ab.done.Wait()`. This will lead to panic.

I know it is a rare case that probably won't happen but it highlights that concurrency is hard.

Reply:
Fixed ✅